### PR TITLE
Support multiple HPO experiments in one process

### PR DESCRIPTION
**Pull request by liuzhe-lz**
### Description ###

Support creating and running multiple HPO experiments in one Python process, either sequentially or concurrently.

Don't want to add test until new IT framework complete.

### Checklist ###
  - [ ] test case
  - ~~doc~~

### How to test ###

```python
exp1 = Experiment(...)
exp1.start()
exp2 = Experiment(...)
exp3 = Experiment(...)
exp2.start()
exp3.start()
wait_completion(exp1)  # unbelievably we do not have this API
exp1.stop()
exp4 = Experiment(...)
exp4.start()
# ctrl-c
```

### CODE REVIEW

File: nni/runtime/protocol.py [hunk 1]
```diff
@@ -7,26 +7,12 @@
...
-    global _use_ws
-    _use_ws = True
-    shim.connect(url)
 
 def send(command: CommandType, data: str) -> None:
-    if _use_ws:
-        shim.send(command, data)
-    else:
-        legacy.send(command, data)
+    legacy.send(command, data)
```
Comment (J-shang) [inline nni/runtime/protocol.py:12]: do we have somewhere still use `protocal.send` or `protocal.receive`?

**Review by J-shang — APPROVED** [review]

File: nni/runtime/protocol.py [hunk 2]
```diff
@@ -7,26 +7,12 @@
 
 from .tuner_command_channel.command_type import CommandType
 from .tuner_command_channel import legacy
-from .tuner_command_channel import shim
```
Comment (matluster) [inline nni/runtime/protocol.py:10]: This file is abandoned?

File: nni/runtime/protocol.py (hunk 1 above)
Comment (liuzhe-lz) [reply nni/runtime/protocol.py:12]: Only in legacy UT.

File: nni/runtime/protocol.py (hunk 2 above)
Comment (liuzhe-lz) [reply nni/runtime/protocol.py:10]: Yes. The migration is nearly complete.

Comment (matluster) [reply nni/runtime/protocol.py:10]: Well, in that case, I suggest adding a deprecated warning when someone tries to import this file.

**Review by ultmaster — APPROVED** [review]

File: nni/runtime/protocol.py (hunk 2 above)
Comment (liuzhe-lz) [reply nni/runtime/protocol.py:10]: Removed.

File: nni/tools/nnictl/launcher.py [hunk 3]
```diff
@@ -88,11 +88,14 @@ def create_experiment(args):
...
 
-    _logger.info(f'To stop experiment run "nnictl stop {exp.id}" or "nnictl stop --all"')
-    _logger.info('Reference: https://nni.readthedocs.io/en/stable/reference/nnictl.html')
+    if foreground:
+        exp.run(port, debug=debug)
+
+    else:
+        exp.start(port, debug, RunMode.Detach)
+        _logger.info(f'To stop experiment run "nnictl stop {exp.id}" or "nnictl stop --all"')
+        _logger.info('Reference: https://nni.readthedocs.io/en/stable/reference/nnictl.html')
```
Comment (QuanluZhang) [inline nni/tools/nnictl/launcher.py:98]: these lines imply that the only different between `exp.start` and `exp.run` is that `exp.run` will wait until experiment completes, right? better to put this information in docstring.

Comment (liuzhe-lz) [reply nni/tools/nnictl/launcher.py:98]: If I remember correctly, end users should always use `run()` according a previous discussion. `run()` has carefully designed and reviewed parameters, while `start()` has kept its "experimental" signature.
We never mention `start()` in tutorial.
I have once thought about renaming `start()` to `_start()`, but gave up due to compatibility concern.

Back to the code. I have to admit line 93 is incorrect. It should be `exp.start(..., Foreground) ; exp.wait_completion()`. I don't know why I forgot about that...

**Pull request merged**